x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java

```diff
@@ -346,4 +419,55 @@ private static String toDocId(String application, String name) {
         return DOC_TYPE_VALUE + "_" + application + ":" + name;
     }
 
+    public void onSecurityIndexStateChange(SecurityIndexManager.State previousState, SecurityIndexManager.State currentState) {
+        if (isMoveFromRedToNonRed(previousState, currentState) || isIndexDeleted(previousState, currentState)
+            || previousState.isIndexUpToDate != currentState.isIndexUpToDate) {
+            invalidateAll();
+        }
+    }
+
+    public void invalidate(Collection<String> updatedApplicationNames) {
+        if (descriptorsCache == null) {
+            return;
+        }
+        logger.debug("Invalidating application privileges caches for: {}", updatedApplicationNames);
+        try (ReleasableLock ignored = invalidationWriteLock.acquire()) {
+            numInvalidation.incrementAndGet();
+            final Set<String> uniqueNames = Set.copyOf(updatedApplicationNames);
+            // Always completely invalidate application names cache due to wildcard
+            applicationNamesCache.invalidateAll();
+            uniqueNames.forEach(descriptorsCache::invalidate);
+        }
```

Comment:
I think we can release the lock immediately after incrementing `numInvalidation`. It will mean that, in theory, we could invalidate things that we don't need to, but would mean holding a lock for less time.

Did you consider the trade-off of how long to lock for _vs_ perfect cache accuracy?

If we keep the lock around all the invalidation process, then I think the calculation of `uniqueNames` should be before the lock is acquired.

Reply:
Thanks @tvernum. Your comment is very insightful. You are right we could minimize the locking time in this case. The possibility of invalidating more than necessary should be very low. The `getPrivileges` thread needs to read the incremented value of `numInvalidation` and perform a search query and all these have to complete before the cache is invalidated. The chance should be extremely low. I have updated the code to minimize the locking for both `invalidate` and `invalidateAll`. 

A more possible scenario is "unnecessarily skipping put result in cache". We only cache when `numInvalidation` does not change. But this value changes for both partially and full cache invalidation. In the case of partial invalidation, the things get invalidated may not be relevant to the things that we want to cache. But the code would just skip caching them regardlessly for simplicity. This however has nothing to do with the locking, i.e. the same situation exists before we added the locking. Overall, combined with how Kibana behaves, I think this is an acceptable trade-off because:
* The chance is still low 
* We always fully invalidate `applicationNamesCache` even for partially invalidation. And we decide to keep it this way for simplicity.

Also moved `descriptorsCache != null` check before the read lock, so we do not just lock and find out there is no cache to use. It is an edge case optimization but anyway it is easy to add.